Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,473 @@
+        }
+
+        let isNaN = match.range(at: REGroups.nan.rawValue)
+        if isNaN.location != NSNotFound {
+            self.value = BSONDecimal128.NAN
+            return
+        }
+
+        let isInfinity = match.range(at: REGroups.infinity.rawValue)
+        if isInfinity.location != NSNotFound {
+            if sign < 0 {
+                self.value = BSONDecimal128.NEG_INFINITY
+                return
+            }
+            self.value = BSONDecimal128.INFINITY
+            return
+        }
+
+        var exponentSign = 1
+        let exponentSignRange = match.range(at: REGroups.exponentSign.rawValue)
+        if exponentSignRange.location != NSNotFound, let range = Range(exponentSignRange, in: data) {
+            exponentSign = String(data[range]) == "-" ? -1 : 1
+        }
+
+        let decimalPartNSRange = match.range(at: REGroups.decimalPart.rawValue)
+        guard decimalPartNSRange.location != NSNotFound,
+            let decimalPartRange = Range(decimalPartNSRange, in: data) else {
+            throw BSONError.InvalidArgumentError(message: "Syntax Error: Missing digits in front of the exponent")
+        }
+        let decimalPart = String(data[decimalPartRange])
+        var digits = try BSONDecimal128.convertToDigitsArray(decimalPart)
+
+        var exponent = 0
+        let exponentPartRange = match.range(at: REGroups.exponentPart.rawValue)
+        if exponentPartRange.location != NSNotFound, let range = Range(exponentPartRange, in: data) {
+            exponent = exponentSign * (Int(data[range]) ?? 0)
+        }
+        if decimalPart.contains(".") {
+            let pointIndex = decimalPart.firstIndex(of: ".") ?? decimalPart.endIndex
+            exponent -= decimalPart.distance(from: pointIndex, to: decimalPart.endIndex) - 1
+            if exponent < BSONDecimal128.EXPONENT_MIN {
+                exponent = BSONDecimal128.EXPONENT_MIN
+            }
+        }
+
+        while exponent > BSONDecimal128.EXPONENT_MAX && digits.count <= BSONDecimal128.PRECISION_DIGITS {
+            // Exponent is too large, try shifting zeros into the coefficient
+            digits.append(0)
+            exponent -= 1
+        }
+
+        while exponent < BSONDecimal128.EXPONENT_MIN && !digits.isEmpty {
+            // Exponent is too small, try taking zeros off the coefficient
+            if digits.count == 1 && digits[0] == 0 {
+                exponent = BSONDecimal128.EXPONENT_MIN
+                break
+            }
+
+            if digits.last == 0 {
+                digits.removeLast()
+                exponent += 1
+                continue
+            }
+
+            if digits.last != 0 {
+                // We don't end in a zero and our exponent is too small
+                throw BSONError.InvalidArgumentError(message: "Underflow Error: Value too small")
+            }
+        }
+
+        guard (exponent >= BSONDecimal128.EXPONENT_MIN) && (exponent <= BSONDecimal128.EXPONENT_MAX) else {
```

Comment:
it might be clearer to say the following (up to you though):
```swift
guard BSONDecmial128.EXPONENT_MIN...BSONDecimal128.EXPONENT_MAX.contains(exponent) else { ... }
```